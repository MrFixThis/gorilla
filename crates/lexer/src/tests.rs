use std::collections::VecDeque;

use super::*;

macro_rules! snapshot {
    ( $func:ident, $path:expr ) => {
        #[test]
        fn $func() {
            let src = include_str!($path);
            let mut conf = ::insta::Settings::clone_current();
            conf.set_snapshot_path("../testdata/output/");
            conf.bind(|| ::insta::assert_snapshot!(prepare_snapshot(src)));
        }
    };
}

fn prepare_snapshot(src: &str) -> String {
    let lexer = Lexer::new(src);
    let mut toks: VecDeque<Token> = VecDeque::new();

    loop {
        let tok = lexer.next_tok();
        if tok.kind == Eof {
            break toks.push_back(tok);
        }
        toks.push_back(tok);
    }

    let mut out = String::new();
    for (row, line) in src.lines().enumerate() {
        out += line;
        out += "\n";

        while let Some(tok) = toks.pop_front() {
            if tok.span.start_row != row && tok.kind != Eof {
                toks.push_front(tok);
                break;
            }

            out += &" ".repeat(tok.span.start_col);
            out += &"^".repeat((tok.span.end_col - tok.span.start_col) + 1);
            out += &format!(" {tok:?}\n");
        }
    }

    out
}

snapshot!(test_code_snipet_tokenization, "../testdata/snapshots/code_snipet.gl");
snapshot!(test_comment_tokenization, "../testdata/snapshots/comment.gl");
snapshot!(test_container_types_tokenization, "../testdata/snapshots/container_type.gl");
snapshot!(test_literal_error_tokenization, "../testdata/snapshots/literal_error.gl");
snapshot!(test_function_tokenization, "../testdata/snapshots/function.gl");
snapshot!(test_identifier_tokenization, "../testdata/snapshots/identifier.gl");
snapshot!(test_literal_tokenization, "../testdata/snapshots/literal.gl");
