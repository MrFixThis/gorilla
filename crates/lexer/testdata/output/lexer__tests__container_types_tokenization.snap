---
source: crates/lexer/src/tests.rs
expression: prepare_snapshot(src)
---
var my_array = [1, 2, 3, 4, 5];
^^^ Keyword(var) -> [0 - 0, 0 - 2]
    ^^^^^^^^ Ident(my_array) -> [0 - 4, 0 - 11]
             ^ Eq -> [0 - 13, 0 - 13]
               ^ OpenDel(Bracket) -> [0 - 15, 0 - 15]
                ^ Literal(Int, 1) -> [0 - 16, 0 - 16]
                 ^ Comma -> [0 - 17, 0 - 17]
                   ^ Literal(Int, 2) -> [0 - 19, 0 - 19]
                    ^ Comma -> [0 - 20, 0 - 20]
                      ^ Literal(Int, 3) -> [0 - 22, 0 - 22]
                       ^ Comma -> [0 - 23, 0 - 23]
                         ^ Literal(Int, 4) -> [0 - 25, 0 - 25]
                          ^ Comma -> [0 - 26, 0 - 26]
                            ^ Literal(Int, 5) -> [0 - 28, 0 - 28]
                             ^ CloseDel(Bracket) -> [0 - 29, 0 - 29]
                              ^ Semicolon -> [0 - 30, 0 - 30]
var my_tuple = (1, 4, 8);
^^^ Keyword(var) -> [1 - 0, 1 - 2]
    ^^^^^^^^ Ident(my_tuple) -> [1 - 4, 1 - 11]
             ^ Eq -> [1 - 13, 1 - 13]
               ^ OpenDel(Parenthesis) -> [1 - 15, 1 - 15]
                ^ Literal(Int, 1) -> [1 - 16, 1 - 16]
                 ^ Comma -> [1 - 17, 1 - 17]
                   ^ Literal(Int, 4) -> [1 - 19, 1 - 19]
                    ^ Comma -> [1 - 20, 1 - 20]
                      ^ Literal(Int, 8) -> [1 - 22, 1 - 22]
                       ^ CloseDel(Parenthesis) -> [1 - 23, 1 - 23]
                        ^ Semicolon -> [1 - 24, 1 - 24]
var my_map = { "nice": 5, "lang": 10 };
^^^ Keyword(var) -> [2 - 0, 2 - 2]
    ^^^^^^ Ident(my_map) -> [2 - 4, 2 - 9]
           ^ Eq -> [2 - 11, 2 - 11]
             ^ OpenDel(Brace) -> [2 - 13, 2 - 13]
               ^^^^^ Literal(Str, "nice") -> [2 - 15, 2 - 19]
                     ^ Colon -> [2 - 21, 2 - 21]
                       ^ Literal(Int, 5) -> [2 - 23, 2 - 23]
                        ^ Comma -> [2 - 24, 2 - 24]
                          ^^^^^ Literal(Str, "lang") -> [2 - 26, 2 - 30]
                                ^ Colon -> [2 - 32, 2 - 32]
                                  ^^ Literal(Int, 10) -> [2 - 34, 2 - 35]
                                     ^ CloseDel(Brace) -> [2 - 37, 2 - 37]
                                      ^ Semicolon -> [2 - 38, 2 - 38]
^ Eof -> [0 - 0, 0 - 0]

